lib/internal/modules/esm/loader.js

```diff
@@ -194,13 +220,68 @@ class ESMLoader {
         this.#loaders,
         FunctionPrototypeBind(loader, null), // [1]
       );
+      if (getFileSystem) ArrayPrototypePush(
+        this.#fileSystemBuilders,
+        FunctionPrototypeBind(getFileSystem, null), // [1]
+      );
     }
 
     // [1] ensure hook function is not bound to ESMLoader instance
 
+    this.buildFileSystem();
     this.preload();
   }
 
+  buildFileSystem() {
+    // Note: makes assumptions as to how chaining will work to demonstrate
+    // the capability; subject to change once chaining's API is finalized.
+    const fileSystemFactories = ArrayPrototypeSlice(this.#fileSystemBuilders);
+
+    const defaultFileSystemFactory = fileSystemFactories[0];
+    let finalFileSystem =
+      defaultFileSystemFactory();
+
+    const asyncKeys = ArrayPrototypeFilter(
+      ObjectKeys(finalFileSystem),
+      (name) => !StringPrototypeEndsWith(name, 'Sync'),
+    );
+
+    for (let i = 1; i < fileSystemFactories.length; ++i) {
+      const currentFileSystem = finalFileSystem;
+      const fileSystem = fileSystemFactories[i](() => currentFileSystem);
+
+      // If the loader specifies a sync hook but omits the async one we
+      // leverage the sync version by default, so that most hook authors
+      // don't have to write their implementations twice.
+      for (let j = 0; j < asyncKeys.length; ++j) {
+        const asyncKey = asyncKeys[j];
+        const syncKey = `${asyncKey}Sync`;
+
+        const hasAsync = ObjectPrototypeHasOwnProperty(fileSystem, asyncKey);
+        const hasSync = ObjectPrototypeHasOwnProperty(fileSystem, syncKey);
+
+        if (
+          !hasAsync &&
+          hasSync
+        ) {
+          fileSystem[asyncKey] = async (...args) => {
+            return fileSystem[syncKey](...args);
+          };
```

Comment:
Imo it'd be best to attempt to keep a consistent API - otherwise loader authors would be effectively forced to do:

```ts
const content = await Promise.resolve(fs.readFile(path));
```

The extra `Promise.resolve` wrapper wouldn't look very idiomatic.

Reply:
Another solution is perhaps to just drop this "automatically bind sync functions into async slots", since it comes with its own drawbacks (it can be seen as a footgun, since it makes all async filesystem operations silently turn sync).

I'm not that attached to it, so dropping it would be fine by me.

Reply:
> otherwise loader authors would be effectively forced to do:

I don't think that this is correct, you can await non-Promise objects:

```js
console.log({}); // {}
console.log(await {}); // {}
console.log(await Promise.resolve({})); // {}
```

I don't feel strongly about this, but imho it wouldn't be to bad as an API.

Reply:
Yes: being able to `await` a non-promise was explicitly for this scenario (can find the citation if needed).